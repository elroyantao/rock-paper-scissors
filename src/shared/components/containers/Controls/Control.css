.Controls {
  display: flex;
  flex-direction: column;
  align-items: center;
  margin-top: 40px;
}

.Controls-result {
  padding: 20px;
  font-size: 18px;
}

.Controls-result--win {
  composes: Controls-result;
  color: #59d482;
}
.Controls-result--loose {
  composes: Controls-result;
  color: #ec6d6d;
}
.Controls-result--tie {
  composes: Controls-result;
  color: #f1be35;
}


.Controls-playAgain {
  padding: 5px 20px;
  background-color: #A5E3F4;
  cursor: pointer;
  border: 1px solid gray;
  border-radius: 5px;
}
