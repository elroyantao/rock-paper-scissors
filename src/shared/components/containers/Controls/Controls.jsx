import React, { Component } from 'react'
import PropTypes from 'prop-types'
import { connect } from 'react-redux'

import InProgress from '../../presentation/InProgress/InProgress'
import Hands from '../../presentation/Hands/Hands'
import { applyHand } from '../../../actions/rpslsAction'
import { clearHand } from '../../../actions/handAction'
import styles from './Control.css'

@connect((state) => ({
  result: state.hand.result,
  playerHand: state.hand.playerHand,
  computerHand: state.hand.computerHand
}), { applyHand, clearHand })
export default class Controls extends Component {
  static propTypes = {
    result: PropTypes.string,
    playerHand: PropTypes.number,
    computerHand: PropTypes.number,
    applyHand: PropTypes.func,
    clearHand: PropTypes.func
  }

  renderResult() {
    const { result } = this.props
    if (result === '') return null
    let message
    switch (result) {
      case 'win':
        message = 'You win!!'
        break
      case 'loose':
        message = 'Computer wins!!'
        break
      case 'tie':
        message = 'It\'s a tie!!'
        break
      default:
        message = ''
    }
    return (
      <div className={styles[`Controls-result--${result}`]}>
        {message}
      </div>
    )
  }

  renderPlayAgainButton() {
    const { playerHand, computerHand, clearHand } = this.props
    if (playerHand === null || computerHand === null) return null
    return (
      <div >
        <button className={styles['Controls-playAgain']} onClick={clearHand} >Play Again</button>
      </div>
    )
  }

  render() {
    const { applyHand } = this.props
    return (
      <div className={styles.Controls}>
        <Hands applyHand={applyHand} />
        <InProgress />
        {this.renderResult()}
        {this.renderPlayAgainButton()}
      </div>
    )
  }
}
