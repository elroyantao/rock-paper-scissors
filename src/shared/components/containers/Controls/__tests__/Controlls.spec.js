import testComponentHelper from '../../../../lib/test-helper'
import Controls from '../Controls'


describe('<Controls /> ', () => {
  const initialState = {
    result: '',
    playerHand: null,
    computerHand: null,
    applyHand: jest.fn(),
    clearHand: jest.fn()
  }
  const renderComponent = testComponentHelper(Controls.WrappedComponent)
  describe('@renders', () => {
    it('in default state', () => {
      expect(renderComponent(initialState).getTree())
        .toMatchSnapshot()
    })
    it('when result is win', () => {
      expect(renderComponent({
        ...initialState,
        result: 'win'
      }).getTree())
        .toMatchSnapshot()
    })
    it('when result is loose', () => {
      expect(renderComponent({
        ...initialState,
        result: 'loose'
      }).getTree())
        .toMatchSnapshot()
    })
    it('when result is tie', () => {
      expect(renderComponent({
        ...initialState,
        result: 'tie'
      }).getTree())
        .toMatchSnapshot()
    })
    it('when playerHand and computerHand is not null it should show the play again button', () => {
      expect(renderComponent({
        ...initialState,
        playerHand: 3,
        computerHand: 2
      }).getTree())
        .toMatchSnapshot()
    })
  })
  describe('@events', () => {
    it('should call clearHand on click of play again', () => {
      const { wrapper, instance } = renderComponent({
        ...initialState,
        playerHand: 3,
        computerHand: 2
      })
      expect(instance.props.clearHand).not.toHaveBeenCalled()
      wrapper.find('.Controls-playAgain').simulate('click')
      expect(instance.props.clearHand).toHaveBeenCalledTimes(1)
    })
  })
})
