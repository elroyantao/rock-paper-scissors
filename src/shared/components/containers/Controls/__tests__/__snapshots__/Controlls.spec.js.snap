// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Controls />  @renders in default state 1`] = `
<div
  className="Controls"
>
  <Connect(Controls)
    applyHand={[Function]}
  />
  <Connect(InProgress) />
</div>
`;

exports[`<Controls />  @renders when playerHand and computerHand is not null it should show the play again button 1`] = `
<div
  className="Controls"
>
  <Connect(Controls)
    applyHand={[Function]}
  />
  <Connect(InProgress) />
  <div>
    <button
      className="Controls-playAgain"
      onClick={[Function]}
    >
      Play Again
    </button>
  </div>
</div>
`;

exports[`<Controls />  @renders when result is loose 1`] = `
<div
  className="Controls"
>
  <Connect(Controls)
    applyHand={[Function]}
  />
  <Connect(InProgress) />
  <div
    className="Controls-result--loose"
  >
    Computer wins!!
  </div>
</div>
`;

exports[`<Controls />  @renders when result is tie 1`] = `
<div
  className="Controls"
>
  <Connect(Controls)
    applyHand={[Function]}
  />
  <Connect(InProgress) />
  <div
    className="Controls-result--tie"
  >
    It's a tie!!
  </div>
</div>
`;

exports[`<Controls />  @renders when result is win 1`] = `
<div
  className="Controls"
>
  <Connect(Controls)
    applyHand={[Function]}
  />
  <Connect(InProgress) />
  <div
    className="Controls-result--win"
  >
    You win!!
  </div>
</div>
`;
